### PR TITLE
8.0 sale recalculation optionnaly without name

### PR DESCRIPTION
**Pull request by petrus-v**
This is a purpose to improve `sale_order_price_recalculation` module:
- [x] The existing button is reseting price only (without changing descriptions)
- [x] Adding  a second button to reset sale order lines descriptions only 
- [x] hide those button links when it's a quotation (sale order is draft)
- [x] Move buttons below sale order lines

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Good addition, thanks!

:+1:

**Comment by qtheuret** [comment]
Thanks for this improvement :+1:

**Comment by rafaelbn** [comment]
I have tested in runbot and it looks fine but when duplicating Sale Order:

```
Traceback (most recent call last):
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/http.py", line 537, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/http.py", line 574, in dispatch
    result = self._call_function(**self.params)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/http.py", line 310, in _call_function
    return checked_call(self.db, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/service/model.py", line 113, in wrapper
    return f(dbname, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/http.py", line 307, in checked_call
    return self.endpoint(*a, **kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/http.py", line 803, in __call__
    return self.method(*args, **kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/http.py", line 403, in response_wrap
    response = f(*args, **kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/addons/web/controllers/main.py", line 948, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/addons/web/controllers/main.py", line 940, in _call_kw
    return checked_call(request.db, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/service/model.py", line 113, in wrapper
    return f(dbname, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/addons/web/controllers/main.py", line 939, in checked_call
    return getattr(request.registry.get(model), method)(request.cr, request.uid, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/api.py", line 241, in wrapper
    return old_api(self, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/api.py", line 403, in old_api
    result = new_api(recs, *args, **kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/api.py", line 407, in new_api
    result = [method(rec, *args, **kwargs) for rec in self]
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/addons/sale_quotation_number/model/sale_order.py", line 33, in copy
    return super(SaleOrder, self).copy(default=default)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/api.py", line 239, in wrapper
    return new_api(self, *args, **kwargs)
TypeError: copy() got an unexpected keyword argument 'default'
```

What this PR does?
- If you add a product in a SO and change Description and price we find 2 "links" (not buttons)


- If you click on "Recalculates prices only" Unit price in sale order lines get default price from product
- If you click on "Recalculate prices and reset descriptions", Unit price and descriptions in sale order lines get default values from product

Please @petrus-v  could you explain me please when is this needed? I try to imagine but Odoo changes default Unit price to default just changing Quantity (this is very quick) and you want if you have already description changed I can imagen that you need the default one when duplicagin SO. Isn't it?

Thanks in advance

**Comment by yourix** [comment]
On Tue, Sep 22, 2015 at 2:46 AM, Rafael Blasco notifications@github.com
wrote:

> I have tested in runbot and it looks fine but when duplicating Sale Order:
> 
> Traceback (most recent call last):
>   File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/http.py", line 537, in _handle_exception
>     return super(JsonRequest, self)._handle_exception(exception)
>   File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/http.py", line 574, in dispatch
>     result = self._call_function(*_self.params)
>   File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/http.py", line 310, in _call_function
>     return checked_call(self.db, *args, *_kwargs)
>   File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/service/model.py", line 113, in wrapper
>     return f(dbname, _args, *_kwargs)
>   File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/http.py", line 307, in checked_call
>     return self.endpoint(_a, *_kw)
>   File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/http.py", line 803, in __call__
>     return self.method(_args, *_kw)
>   File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/http.py", line 403, in response_wrap
>     response = f(_args, *_kw)
>   File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/addons/web/controllers/main.py", line 948, in call_kw
>     return self._call_kw(model, method, args, kwargs)
>   File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/addons/web/controllers/main.py", line 940, in _call_kw
>     return checked_call(request.db, _args, *_kwargs)
>   File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/service/model.py", line 113, in wrapper
>     return f(dbname, _args, *_kwargs)
>   File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/addons/web/controllers/main.py", line 939, in checked_call
>     return getattr(request.registry.get(model), method)(request.cr, request.uid, _args, *_kwargs)
>   File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/api.py", line 241, in wrapper
>     return old_api(self, _args, *_kwargs)
>   File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/api.py", line 403, in old_api
>     result = new_api(recs, _args, *_kwargs)
>   File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/api.py", line 407, in new_api
>     result = [method(rec, _args, *_kwargs) for rec in self]
>   File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/addons/sale_quotation_number/model/sale_order.py", line 33, in copy
>     return super(SaleOrder, self).copy(default=default)
>   File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3115451-214-80fbdc/openerp/api.py", line 239, in wrapper
>     return new_api(self, _args, *_kwargs)
> TypeError: copy() got an unexpected keyword argument 'default'
> 
> What this PR does?
> - If you add a product in a SO and change Description and price we
>   find 2 "links" (not buttons)
> 
> [image: 2015-09-22_2-37-08]
> https://cloud.githubusercontent.com/assets/8736623/10008755/1cff82b4-60d3-11e5-8f9e-b06ab19d5296.png
> - If you click on "Recalculates prices only" Unit price in sale order
>   lines get default price from product
> - If you click on "Recalculate prices and reset descriptions", Unit
>   price and descriptions in sale order lines get default values from product
> 
> Please @petrus-v https://github.com/petrus-v could you explain me
> please when is this needed? I try to imagine but Odoo changes default Unit
> price to default just changing Quantity (this is very quick) and you want
> if you have already description changed I can imagen that you need the
> default one when duplicagin SO. Isn't it?
> 
> Thanks in advance
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/sale-workflow/pull/214#issuecomment-142148225.
> 
> Hi rafael

This is a case needed when I have an old sale order and use prices lists
(based on purchase prices and different margin for my customers).
This features is very usefull to let the sale representative get an old
sale order for one customer, duplicate it (for another customer) , reset
the prices and keep every "custom" descriptions.
It let him be sure than prices are ok without retype every special
descriptions (very usefull for services lines).
Another case (more frequently) is the same sale order for one customer
updated several times per year : the price update will be done in 2 clicks
: duplicate / recalculate without reset description , done.

Thank you

## 

Youri

**Comment by petrus-v** [comment]
@rafaelbn, @yourix , @pedrobaeza , @qtheuret 

Should we display those buttons links only when it's a quotation (when the  sale order state is 'draft')? It may be dangerous to change the price afterwards, what do you think about ?

@rafaelbn 

Thanks for reviews, I think the duplicate error you get is not related to this PR:
- As we can see in Traceback it never through `sale_order_price_recalculation` module.
- This module do nothing about the copy nor this PR
- I can't reproduce it on my local DB with this module alone
- Also I reproduce it on 8.0 runbot branch

I use the "button" term as it was formerly used in the module [description](https://github.com/OCA/sale-workflow/tree/8.0/sale_order_price_recalculation#recalculation-of-sales-order-lines-prices-on-demand)

This PR add only "Recalculates prices only" link button feature, @yourix explain above why this PR. I'm going to push an small update about usage part to explain the differences between those two possibilities.

Regards,
Pierre

**Comment by yourix** [comment]
On Wed, Sep 23, 2015 at 10:14 AM, Pierre Verkest notifications@github.com
wrote:

> @rafaelbn https://github.com/rafaelbn, @yourix
> https://github.com/yourix , @pedrobaeza https://github.com/pedrobaeza
> , @qtheuret https://github.com/qtheuret
> 
> Should we display those buttons links only when it's a quotation (when the
> sale order state is 'draft')? It may be dangerous to change the price
> afterwards, what do you think about ?
> 
> Hello,

In my opinion, those recalculation buttons are only necessary in draft
state (and dangerous in other states),

## Regards,

YP

**Comment by petrus-v** [comment]
I update the PR description according the last commit.

**Comment by ccomb** [comment]
From an ergonomic point of view I have two remarks:
- I find it weird that the buttons be located in the middle of the top fields, for something that modify the sale order lines. IMO they should be close to the order lines, maybe below the m2m field.
- I think the purpose of the buttons should be clearly splitted in two different non-overlapping actions : 1 button for resetting the prices, and 1 button for resetting the descriptions. The button label would be shorter and it would allow to only reset the descriptions.

**Comment by pedrobaeza** [comment]
For the first point, the button was placed there (by me), because is next to the pricelist field, that it's the one that mainly change the prices, but it can be also where you say.

About the second, it can be also another approach. I don't have any preference in this case.

File: sale_order_price_recalculation/i18n/fr.po [hunk 1]
```diff
@@ -1,26 +1,33 @@
...
+"Content-Transfer-Encoding: 8bit\n"
 "Plural-Forms: nplurals=2; plural=(n > 1);\n"
+"X-Generator: Poedit 1.6.10\n"
 
 #. module: sale_order_price_recalculation
 #: view:sale.order:sale_order_price_recalculation.sorp_view_order_form
-msgid "(Recalculate prices)"
-msgstr ""
+msgid "(Recalculate prices and reset descriptions)"
+msgstr "(Recalcule des prix et descriptions)"
```
Comment (ccomb) [inline sale_order_price_recalculation/i18n/fr.po]: Recalcule → Recalcul

File: sale_order_price_recalculation/i18n/fr.po [hunk 2]
```diff
@@ -1,26 +1,33 @@
...
+
+#. module: sale_order_price_recalculation
+#: view:sale.order:sale_order_price_recalculation.sorp_view_order_form
+msgid "(Recalculate prices only)"
+msgstr "(Recalcule des prix seulement)"
```
Comment (ccomb) [inline sale_order_price_recalculation/i18n/fr.po]: Recalcule → Recalcul

**Comment by petrus-v** [comment]
As asked,

I split functionalities and move buttons bellow sale order lines.

**Comment by pedrobaeza** [comment]
:+1:

**Comment by ccomb** [comment]
:+1:

**Pull request merged**